src/client/opamPinCommand.ml

```diff
@@ -75,15 +75,31 @@ let get_source_definition ?version ?subpath ?locked st nv url =
     opam
   in
   let open OpamProcess.Job.Op in
-  OpamUpdate.fetch_dev_package url srcdir ?subpath nv @@| function
+  OpamUpdate.fetch_dev_package url internal_pindir ?subpath nv @@+ function
   | Not_available (_,s) -> raise (Fetch_Fail s)
   | Up_to_date _ | Result _ ->
-    let srcdir =
+    let opam_dir_lookup () =
       let u = OpamFile.URL.url url in
       match OpamUrl.local_dir u, u.OpamUrl.backend with
-      | Some dir, #OpamUrl.version_control -> dir
-      | _, _ -> srcdir
+      | Some dir, #OpamUrl.version_control ->
+        let get_branch d =
+          let url = OpamUrl.of_string (OpamFilename.Dir.to_string d) in
+          OpamRepository.revision d
+            { url with
+              OpamUrl.transport = u.transport;
+              backend = u.backend}
+        in
+        get_branch dir @@+ fun local_branch ->
+        get_branch internal_pindir @@| fun distant_branch ->
+        (match local_branch, distant_branch with
+         | Some l, Some d when OpamPackage.Version.equal l d -> dir
```

Comment:
i'm not sure why `OpamRepository.revision` returns a version type but this might be something we might want to change to `string`. It is rather confusing to see it used like a version number.